[#3] 회원가입 기능 구현

ethan-sh-lee opened the pull request:
마이바티스를 이용하여 로컬 컴퓨터에 mysql과 연동하여 회원가입 기능을 구현하였습니다.

ethan-sh-lee committed:
데이터베이스 설정 변경

ethan-sh-lee committed:
마이바티스 설정 추가

ethan-sh-lee committed:
마이바티스 매퍼파일 추가

ethan-sh-lee committed:
마이바티스 매퍼 인터페이스 추가

ethan-sh-lee committed:
판매자 회원가입 기능 추가

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerController.java:
```diff
@@ -0,0 +1,52 @@
+
+
+@RestController
+@RequestMapping("/customer")
```
여기는 아래 문서를 읽어보시고 그에 맞춰서 고쳐보시면 좋을 것 같네요~

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerController.java:
```diff
@@ -0,0 +1,52 @@
+@RequestMapping("/customer")
+public class CustomerController {
+
+  @Autowired
```
일반적으로는 생성자 주입을 권장합니다~ 왜그럴까요?

ethan-sh-lee replied:
1. 생성자 주입은 객체가 생성될 때 주입이 되는 것을 확신할 수 있어 null일 확률이 없습니다. setter와 field 주입은 null일 확률이 존재합니다.

2. 순환 참조 문제를 컴파일 타임에 방지할 수 있습니다.

3. 생성자 주입은 객체 주입 후 변경될 수 없기 때문에 객체가 불변하여 멀티쓰레드 환경에서 개발자가 의도치 않은 객체의 변경을 방지할 수 있습니다.

4. 생성자 주입은 스프링 4.3 이후 애노테이션을 입력하지 않아도 생성자 주입이 가능하여 스프링 의존성이 적어 단위 테스트시 용이합니다.

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerController.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  @GetMapping("/{id}")
+  public CustomerDTO getCustomerDetail(@PathVariable String id){
```
id가 숫자가 아니라 문자열인 이유가 무엇일까요?

ethan-sh-lee replied:
id는 로그인을 위한 회원 email을 의미하기 때문입니다.
맨 처음에 id는 pimary key로써 email과 함께 존재하였습니다.
회의 후 id값은 회원 식별 용도로써만 존재하고 email값이 이를 대체할 수 있다고 생각하여 email 컬럼을 지우고 id 컬럼에 email 값을 넣기로 하였습니다.

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerController.java:
```diff
@@ -0,0 +1,52 @@
+    return customerService.getCustomerDetailById(id);
+  }
+
+  @PutMapping("/pwd/{id}")
```
위에 달아드린 링크를 보고 URL을 전반적으로 한번 설계해보시면 좋을 것 같아요~

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerController.java:
```diff
@@ -0,0 +1,52 @@
+  }
+
+  @PutMapping("/pwd/{id}")
+  public boolean modifyCustomerPwd(@PathVariable String id, @RequestBody CustomerDTO customerDTO){
```
성공 여부를 리턴하는 것 같은데 성공여부는 굳이 값으로 리턴하지 않아도 괜찮습니다~
성공하지 못했을 때에는 예외를 던지는게 낫기 때문입니다

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerDTO.java:
```diff
@@ -0,0 +1,20 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Getter
+@Setter
```
이 Setter를 달아주면 immutable이 깨지게 됩니다. 깨지면 뭐가 문제일까요?

ethan-sh-lee replied:
객체 내부 상태를 수정할 수 있어 데이터가 언제든 변할 수 있어 신뢰할 수 없습니다. 또한 멀티 쓰레드 환경에서 thread-safe하지 않습니다.

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerMapper.java:
```diff
@@ -0,0 +1,17 @@
+  int updateCustomerName(String id, String updateName);
+  int updateCustomerPhoneNum(String id, String updatePhoneNum);
+  int deleteCustomerById(String id);
+
+
```
여기는 필요없는 공백 같습니다~

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerMapper.java:
```diff
@@ -0,0 +1,17 @@
+public interface CustomerMapper {
+
+  CustomerDTO selectCustomerDetailById(String id);
+  void insertCustomer(CustomerDTO params) throws SQLIntegrityConstraintViolationException;
```
메소드 사이에 개행을 하나씩 넣어주면 가독성이 향상될 것 같네요~

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,52 @@
+    try{
+      customerMapper.insertCustomer(params);
+    }catch (Exception e){
+      e.printStackTrace();
```
이 메소드는 절대 쓰면 안돼는 메소드입니다. 왜일까요?

ethan-sh-lee replied:
printStackTrace는 정보를 출력하는 메소드에 sychronized 블록으로 감싸져 있어 멀티쓰레드 환경에서 성능 저하가 발생됩니다. 또한 필요 없는 정보를 취합하여 출력하기 때문에 서버에 의미없는 부하를 줍니다.

ethan-sh-lee committed:
DI 방식 변경

field 의존성 주입을 constroctor 의존성 주입으로 변경

isaac1102 commented on src/main/java/com/project/smartStore/customer/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,52 @@
+  @Override
+  public boolean registerCustomer(CustomerDTO params){
+    try{
+      customerMapper.insertCustomer(params);
```
비밀번호 암호화 과정 없이 평문으로 저장되는 방식인가요?

ethan-sh-lee replied:
넵 우선 암호화 없이 구현하였고 추후에 추가할 예정입니다!

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerController.java:
```diff
@@ -10,14 +9,15 @@
-  private CustomerService customerService;
+  private final CustomerService customerService;
 
+  public CustomerController(CustomerService customerService) {
```
이거는 lombok으로 간소화가 가능합니다~

ethan-sh-lee committed:
DI 구현 방식 변경

생성자를 직접 구현한 것에서 lombok 애노테이션 방식으로 변경

ethan-sh-lee committed:
가독성 향상을 위한 코드 개행 수정

ethan-sh-lee committed:
필요없는 리턴타입을 void로 수정

CustomerController, CustomerMapper, CustomerService, CustomerServiceImpl에서 성공 여부를 반환하는 로직을 제거하고 성공시 반환하지 않고 실패시에만 예외처리하는 방식으로 수정

ethan-sh-lee committed:
필요없는 lombok 애노테이션 삭제

@setter를 삭제하여 객체의 무분별한 수정을 방지
사용하지 않는 @ToString을 삭제

f-lab-dev commented:
리뷰받은 부분 중 남은 것들도 수정해주시고 질문드린것도 답변해주시면 될 것 같습니다~

ethan-sh-lee committed:
CumstomerController rest api 수정

customer를 customers로 수정
customers 회원 정보의 부분 수정을 위해 put을 patch로 수정

ethan-sh-lee committed:
try-catch문 제거

관심사를 분리하기 위해 개별적인 try-catch문은 지양하고 exception handler로 구현 예정

ethan-sh-lee committed:
CustomerServiceImpl DI 방식 수정

CustomMapper를 주입할 때 필드 주입 방식에서 생성자 주입 방식으로 수정

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerController.java:
```diff
@@ -27,20 +27,17 @@ public CustomerDTO getCustomerDetail(@PathVariable String id) {
-    customerService.modifiyPhoneNum(id, customerDTO.getPhoneNum());
+  @PatchMapping("/{id}")
+  public void modifyCustomer(@PathVariable String id, @RequestBody CustomerDTO customerDTO) {
+    if (customerDTO.getName() != null) {
+      customerService.modifiyName(id, customerDTO.getName());
+    }
+    if (customerDTO.getPassword() != null) {
+      customerService.modifiyPwd(id, customerDTO.getPassword());
+    }
+    if (customerDTO.getPhoneNum() != null) {
+      customerService.modifiyPhoneNum(id, customerDTO.getPhoneNum());
+    }
```
너무 한번에 많은것을 변경하는 것 같은데 이거는 PATCH가 아니라 PUT 아닐까요?
또한 쿼리가 한번에 최대 3개까지 나갈 수 있다는점도 성능에 부하가 있을 것 같아요~

f-lab-dev reviewed (commented):
암호화도 여기서 같이 구현해주시는게 나을 것 같아요~
암호화를 미리 해두지 않으면 DB에 저장될때 싱크가 안맞을 수도 있을 것 같아요

ethan-sh-lee committed:
비밀번호 암호화 구현

f-lab-dev commented on src/main/java/com/project/smartStore/customer/CustomerDTO.java:
```diff
@@ -3,13 +3,15 @@
 public class CustomerDTO {
 
   private String id;
+  @Setter
```
굳이 여기에 setter를 붙이기보단 객체를 하나 더 만들어주는 방식으로 가야 immutable을 지킬 수 있을 것 같습니다~

ethan-sh-lee replied:
기존에 비밀번호 암호화를 한 후 setter를 이용하여 변경하였는데 다시 객체를 생성하여 암호화된 비밀번호를 담을 수 있도록 수정하겠습니다.

f-lab-dev commented on src/main/java/com/project/smartStore/security/EncryptionConverter.java:
```diff
@@ -0,0 +1,37 @@
+  @Value("${enc.salt}")
+  public String salt;
+
+  public String ConvertSHA256WithSalt(String inputString) throws NoSuchAlgorithmException {
```
굳이 checked exception을 던질 필요가 있을까요?

ethan-sh-lee replied:
예외가 발생하여도 당장 처리하기 어려운 예외는 runtime exception으로 바꾸도록 처리하는 것이 좋을 것 같습니다.

ethan-sh-lee committed:
exeception 처리방식 수정

예외 발생시 처리할 수 없는 예외는 runtime exception으로 전환

ethan-sh-lee committed:
customerDTO setter 삭제

customerDTO의 immutable한 객체로 만들기 위해 setter를 삭제

ethan-sh-lee committed:
customerDTO setter 대신 객체 생성으로 수정

customerDTO의 password를 암호화한 것을 setter로 수정하는 대신 새 객체를 만들어 값을 넣어주었다.

ethan-sh-lee committed:
기능 : customer 정보 수정 기능 변경

기존에 patch 메소드이고 password, name, phoneNum 세개의 변경 쿼리를 따로 두었는데 하나의 메소드에서 3개의 쿼리를 날릴 수 있기 때문에 성능 저하를 가져올 수 있어 put메소드로 변경하고 하나의 쿼리만 날릴 수 있도록 수정하였다.
